pkg/accounting/mock/accounting.go

```diff
@@ -1,49 +1,140 @@
+	balancesFunc func() (map[string]int64, error)
 }
 
-func (ma *MockAccounting) Reserve(peer swarm.Address, price uint64) error {
-	return nil
+// WithReserveFunc sets the mock Reserve function
+func WithReserveFunc(f func(peer swarm.Address, price uint64) error) Option {
+	return optionFunc(func(s *Service) {
+		s.reserveFunc = f
+	})
+}
+
+// WithReleaseFunc sets the mock Release function
+func WithReleaseFunc(f func(peer swarm.Address, price uint64)) Option {
+	return optionFunc(func(s *Service) {
+		s.releaseFunc = f
+	})
+}
+
+// WithCreditFunc sets the mock Credit function
+func WithCreditFunc(f func(peer swarm.Address, price uint64) error) Option {
+	return optionFunc(func(s *Service) {
+		s.creditFunc = f
+	})
+}
+
+// WithDebitFunc sets the mock Debit function
+func WithDebitFunc(f func(peer swarm.Address, price uint64) error) Option {
+	return optionFunc(func(s *Service) {
+		s.debitFunc = f
+	})
+}
+
+// WithBalanceFunc sets the mock Balance function
+func WithBalanceFunc(f func(swarm.Address) (int64, error)) Option {
+	return optionFunc(func(s *Service) {
+		s.balanceFunc = f
+	})
+}
+
+// WithBalancesFunc sets the mock Balances function
+func WithBalancesFunc(f func() (map[string]int64, error)) Option {
+	return optionFunc(func(s *Service) {
+		s.balancesFunc = f
+	})
 }
 
-func (ma *MockAccounting) Release(peer swarm.Address, price uint64) {
+// NewAccounting creates the mock accounting implementation
+func NewAccounting(opts ...Option) accounting.Interface {
+	mock := new(Service)
+	mock.balances = make(map[string]int64)
+	for _, o := range opts {
+		o.apply(mock)
+	}
+	return mock
+}
 
+// Reserve is the mock function wrapper that calls the set implementation
+func (s *Service) Reserve(peer swarm.Address, price uint64) error {
+	if s.reserveFunc != nil {
+		return s.reserveFunc(peer, price)
+	}
+	return nil
+}
+
+// Release is the mock function wrapper that calls the set implementation
+func (s *Service) Release(peer swarm.Address, price uint64) {
+	if s.releaseFunc != nil {
+		s.releaseFunc(peer, price)
+	}
 }
 
-func (ma *MockAccounting) Credit(peer swarm.Address, price uint64) error {
-	ma.lock.Lock()
-	defer ma.lock.Unlock()
-	ma.balances[peer.String()] -= int64(price)
+// Credit is the mock function wrapper that calls the set implementation
+func (s *Service) Credit(peer swarm.Address, price uint64) error {
+	if s.creditFunc != nil {
+		return s.creditFunc(peer, price)
+	}
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	s.balances[peer.String()] -= int64(price)
 	return nil
 }
 
-func (ma *MockAccounting) Debit(peer swarm.Address, price uint64) error {
-	ma.lock.Lock()
-	defer ma.lock.Unlock()
-	ma.balances[peer.String()] += int64(price)
+// Debit is the mock function wrapper that calls the set implementation
+func (s *Service) Debit(peer swarm.Address, price uint64) error {
+	if s.debitFunc != nil {
+		return s.debitFunc(peer, price)
+	}
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	s.balances[peer.String()] += int64(price)
```

Comment:
would you still not like to increment the changes to balances even though the pluggable function was called? it would be nice to have both working no? maybe this can be in some defer closure which is called at the top of the method (this goes for all of the functions that were modified to accommodate for this change)